src/plugins/reactions/index.js

```diff
@@ -0,0 +1,345 @@
+        });
+    },
+
+    /**
+     * Send a reaction to a message
+     * Implements XEP-0444: Message Reactions
+     * 
+     * @param {Object} message - The message model being reacted to
+     * @param {string} emoji - The emoji reaction (can be unicode or shortname like :joy:)
+     * 
+     * Process:
+     * 1. Convert emoji shortname to unicode if needed
+     * 2. Build XEP-0444 compliant stanza
+     * 3. Send via XMPP connection
+     * 4. Optimistically update local state for immediate UI feedback
+     */
+    sendReaction (message, emoji) {
+        const { $msg } = converse.env;
+        const chatbox = message.collection.chatbox;
+        const msgId = message.get('msgid');
+        const to_jid = chatbox.get('jid');
+        // Default to 'chat' type unless explicitly a groupchat (MUC)
+        const type = chatbox.get('type') === 'groupchat' ? 'groupchat' : 'chat';
+        
+        if (!emoji) return;
+
+        // Convert emoji shortname (e.g. :joy:) to unicode (e.g. 😂)
+        // Check if emoji is already unicode (from emoji picker) or needs conversion (from shortname buttons)
+        let emojiUnicode = emoji;
+        if (emoji.startsWith(':') && emoji.endsWith(':')) {
+            const emojiArray = u.shortnamesToEmojis(emoji, { unicode_only: true });
+            emojiUnicode = Array.isArray(emojiArray) ? emojiArray.join('') : emojiArray;
+        }
+
+        // Build XEP-0444 reaction stanza
+        const reaction = $msg({
+            'to': to_jid,
+            'type': type,
+            'id': u.getUniqueId('reaction')
+        }).c('reaction', {
+            'xmlns': 'urn:xmpp:reactions:0',
+            'id': msgId  // ID of the message being reacted to
+        }).c('emoji').t(emojiUnicode);
+
+        // Send stanza to XMPP server
+        api.send(reaction);
+
+        // Optimistic local update for immediate UI feedback
+        const my_jid = api.connection.get().jid;
+        const currentReactions = message.get('reactions') || {};
+        // Clone to ensure Backbone detects the change
+        const reactions = JSON.parse(JSON.stringify(currentReactions));
+
+        // Remove user's previous reactions (one reaction per message per XEP-0444)
+        for (const existingEmoji in reactions) {
+            const index = reactions[existingEmoji].indexOf(my_jid);
+            if (index !== -1) {
+                reactions[existingEmoji].splice(index, 1);
+                // Clean up emoji key if no users remain
+                if (reactions[existingEmoji].length === 0) {
+                    delete reactions[existingEmoji];
+                }
+            }
+        }
+
+        // Add the new reaction
+        if (!reactions[emojiUnicode]) {
+            reactions[emojiUnicode] = [];
+        }
+        reactions[emojiUnicode].push(my_jid);
+        
+        // Save to model - triggers view re-render
+        message.save({ 'reactions': reactions });
```

Comment:
Here you have a lot of duplicate code.
Moreover, the XEP says that a reaction can be rejected. See https://xmpp.org/extensions/xep-0444.html#rejecting
In such case, you should not show them in the user interface.

So, i think you should just remove this code. In the XMPP world, the "immediate feedback" is not always a good option. You will have safer code if you just rely on the incoming stanzas. (in general, action a user does will induce a stanza from the server, and just handling this stanza in the normal flow always does the trick).